Added anchor support

judax opened the pull request:
Initial implementation for Anchors in ARCore. All the basic structures have been added but there is still testing to be done. Mainly because there seems to be a crash in the event handling.

judax committed:
Added anchor support. There seems to be a crash when the anchorsUpdated
event is fired but the logic is implemented. This commit includes some
debugging logs that should be removed.

Change-Id: I66871218c8ce40dd5c2f07bea6a5f1987e63af11

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+
+const float* Anchor::getModelMatrix() const
+{
+  return (const float*)(&anchorModelMatrix);
```
return glm::value_ptr(anchorModelMatrix);

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
```
auto it = ...

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+    anchors.find(identifier);
+  if (it != anchors.end())
+  {
+    anchors.erase(it);
```
erase actually supports erase-by-key: http://www.cplusplus.com/reference/unordered_map/unordered_map/erase/

This whole method can be (I believe) "anchors.erase(identifier);"

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+{
+  std::vector<std::shared_ptr<Anchor>> updatedAnchors;
+  updatedAnchors.reserve(anchors.size());
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::const_iterator it =
```
auto it

lincolnfrog commented on android_webview/test/shell/tango/jni/TangoHandler.h:
```diff
@@ -77,7 +82,8 @@ class TangoHandler {
 
-  void onCreate(JNIEnv* env, jobject activity, int activityOrientation, int sensorOrientation);
+  void onCreate(JNIEnv* env, jobject activity, 
+  							int activityOrientation, int sensorOrientation);
```
this is all tabs :(

lincolnfrog commented on android_webview/test/shell/tango/jni/TangoHandler.h:
```diff
@@ -89,9 +95,14 @@ class TangoHandler {
   bool hitTest(float x, float y, std::vector<Hit>& hits);
   bool getPlaneDeltas(PlaneDeltas& planeDeltas);
+  std::shared_ptr<Anchor> createAnchor(
+  	const float* anchorModelMatrix);
```
tab

lincolnfrog commented on android_webview/test/shell/tango/jni/TangoHandler.h:
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
+{
+public:
+  virtual void anchorsUpdated(
+  	const std::vector<std::shared_ptr<Anchor>>& anchors) = 0;
```
tab

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
+    const std::vector<float>& modelMatrix) 
+{
+  if (!TangoHandler::getInstance()->isConnected()) 
+  {
```
shouldn't this be on the previous line? I believe that is Google c++ style (I know its js and java style).

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
+  std::shared_ptr<Anchor> anchor = TangoHandler::getInstance()->
+    createAnchor((const float*)(&(modelMatrix[0])));
+  if (!anchor)
+  {
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -309,4 +345,20 @@ void TangoVRDevice::UpdateLayerBounds(mojom::VRLayerBoundsPtr left_bounds,
+  std::vector<mojom::VRAnchorPtr> mojomAnchors;
+  mojomAnchors.resize(anchors.size());
+  for (size_t i = 0; i < anchors.size(); i++)
+  {
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -309,4 +345,20 @@ void TangoVRDevice::UpdateLayerBounds(mojom::VRLayerBoundsPtr left_bounds,
 
+void TangoVRDevice::anchorsUpdated(
+  const std::vector<std::shared_ptr<Anchor>>& anchors)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -269,6 +281,30 @@ mojom::VRPlaneDeltasPtr TangoVRDevice::GetPlaneDeltas() {
 
+mojom::VRAnchorPtr TangoVRDevice::CreateAnchor(
+    const std::vector<float>& modelMatrix) 
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/android/tango/tango_vr_device.cc:
```diff
@@ -239,6 +239,18 @@ static mojom::VRPlanePtr CreateMojomPlane(Plane& plane) {
 }
 
+static mojom::VRAnchorPtr CreateMojomAnchor(const std::shared_ptr<Anchor>& anchor)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+  for (; it != anchors.end(); it++) {
+    std::shared_ptr<Anchor> anchor = it->second;
+    if (anchor->update(newTimestamp, newCameraModelMatrix))
+    {
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+
+std::vector<std::shared_ptr<Anchor>> AnchorManager::update(
+  double newTimestamp, const float* newCameraModelMatrix)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+void AnchorManager::removeAllAnchors()
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
+    anchors.find(identifier);
+  if (it != anchors.end())
+  {
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+void AnchorManager::removeAnchor(uint32_t identifier)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+
+std::shared_ptr<Anchor> AnchorManager::createAnchor(double timestamp,
+  const float* cameraModelMatrix, const float* anchorModelMatrix)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+const float* Anchor::getModelMatrix() const
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+uint32_t Anchor::getIdentifier() const
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+{
+  bool needsUpdate = newTimestamp <= timestamp;
+  if (needsUpdate)
+  {
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+bool Anchor::update(double newTimestamp, const float* newCameraModelMatrix)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on android_webview/test/shell/tango/jni/AnchorManager.cpp:
```diff
@@ -0,0 +1,89 @@
+Anchor::Anchor(double timestamp, const float* cameraModelMatrix,
+               const float* anchorModelMatrix):
+                 identifier(identifierCounter++), timestamp(timestamp)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/vr_device.cc:
```diff
@@ -51,6 +51,15 @@ void VRDevice::OnChanged() {
 }
 
+void VRDevice::OnAnchorsUpdated(std::vector<mojom::VRAnchorPtr> anchors)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on device/vr/vr_device.cc:
```diff
@@ -51,6 +51,15 @@ void VRDevice::OnChanged() {
+  
+  VLOG(0) << "JUDAX: VRDevice::OnAnchorsUpdated -> displays_.size() = " << displays_.size() << ", anchors.size() = " << anchors.size();
+
+  for (const auto& display : displays_)
```
always surround blocks with {} according to Google style

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchor.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace blink {
+
+VRAnchor::VRAnchor()
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchor.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+void VRAnchor::setAnchor(const device::mojom::blink::VRAnchorPtr& anchorPtr)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchor.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+DEFINE_TRACE(VRAnchor)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp:
```diff
@@ -0,0 +1,44 @@
+VRAnchorEvent::VRAnchorEvent() {}
+
+VRAnchorEvent::VRAnchorEvent(const AtomicString& type,
+                           bool canBubble,
```
indentation one more space for this whole block?

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp:
```diff
@@ -0,0 +1,44 @@
+      m_anchors(anchors) {}
+
+VRAnchorEvent::VRAnchorEvent(const AtomicString& type,
+                           const VRAnchorEventInit& initializer)
```
one more space indent

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp:
```diff
@@ -0,0 +1,44 @@
+                           const VRAnchorEventInit& initializer)
+    : Event(type, initializer) {
+  if (initializer.hasDisplay())
+    m_display = initializer.display();
```
always use {} for blocks

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.cpp:
```diff
@@ -0,0 +1,44 @@
+  if (initializer.hasDisplay())
+    m_display = initializer.display();
+
+  if (initializer.hasAnchors())
```
always use {} for blocks

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.h:
```diff
@@ -0,0 +1,57 @@
+
+  ~VRAnchorEvent() override;
+
+  VRDisplay* display() const { return m_display.get(); }
```
I am pretty sure you aren't supposed to have spaces within {} blocks like this according to Google style (at least for python, can't find it for c, but check go/cstyle and see if you can find it).

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRAnchorEvent.h:
```diff
@@ -0,0 +1,57 @@
+ private:
+  VRAnchorEvent();
+  VRAnchorEvent(const AtomicString& type,
+               bool canBubble,
```
one more space indentation for this block

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+VRAnchor* VRDisplay::createAnchor(WTF::Vector<float>& modelMatrix)
+{
+  if (!m_display)
+    return nullptr;
```
always use {} around blocks

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
 }
 
+VRAnchor* VRDisplay::createAnchor(WTF::Vector<float>& modelMatrix)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+}
+
+VRAnchor* VRDisplay::createAnchor(DOMFloat32Array* modelMatrix)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+VRAnchor* VRDisplay::createAnchor(DOMFloat32Array* modelMatrix)
+{
+  if (!m_display)
+    return nullptr;
```
always use {} around blocks

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+}
+
+void VRDisplay::removeAnchor(VRAnchor* anchor)
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+void VRDisplay::removeAnchor(VRAnchor* anchor)
+{
+  if (!m_display)
+    return;
```
always use {} around blocks

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+    // NOTE: Should we compare the pointers directly?
+    if (m_anchors[i]->identifier() == anchor->identifier()) {
+      m_anchors.remove(i);
+      removed = true;
```
any reason to not just break here? you get rid of a bool and an extra check. Are you expecting multiple entries with matching ids?

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -316,6 +318,63 @@ VRPassThroughCamera* VRDisplay::getPassThroughCamera()
+}
+
+HeapVector<Member<VRAnchor>> VRDisplay::getAnchors()
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/WebKit/Source/modules/vr/VRDisplay.cpp:
```diff
@@ -792,6 +851,40 @@ void VRDisplay::OnChanged(device::mojom::blink::VRDisplayInfoPtr display) {
+  //   return;
+  // }
+
+  VLOG(0) << "JUDAX: VRDisplay::OnAnchorsUpdated -> mojomAnchors.size() = " << mojomAnchors.size();
```
do you want to check this in?

lincolnfrog commented on third_party/tango/libtango_chromium/AnchorManager.h:
```diff
@@ -0,0 +1,66 @@
+class AnchorManager;
+
+class Anchor
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/tango/libtango_chromium/AnchorManager.h:
```diff
@@ -0,0 +1,66 @@
+};
+
+class AnchorManager
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/tango/libtango_chromium/AnchorManager.h:
```diff
@@ -0,0 +1,66 @@
+{
+public:
+  std::shared_ptr<Anchor> createAnchor(double timestamp, 
+                                    const float* cameraModelMatrix,
```
add a couple spaces to this block indent

lincolnfrog commented on third_party/tango/libtango_chromium/AnchorManager.h:
```diff
@@ -0,0 +1,66 @@
+    double newTimestamp, const float* newCameraModelMatrix);
+  
+private:
+  std::unordered_map<uint32_t, std::shared_ptr<Anchor>> anchors;
```
m_anchors?

lincolnfrog commented on third_party/tango/libtango_chromium/AnchorManager.h:
```diff
@@ -0,0 +1,66 @@
+  * Return: The list of updated anchors.
+  */
+  std::vector<std::shared_ptr<Anchor>> update(
+    double newTimestamp, const float* newCameraModelMatrix);
```
I think this should be 4 spaces (continuation indent)

lincolnfrog commented on third_party/tango/libtango_chromium/TangoHandler.h:
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
 };
 
+class TangoHandlerEventListener
+{
```
{ should be on the previous line according to Google c++ style (go/cstyle)

lincolnfrog commented on third_party/tango/libtango_chromium/TangoHandler.h:
```diff
@@ -64,6 +61,14 @@ class PlaneDeltas
+{
+public:
+  virtual void anchorsUpdated(
+  	const std::vector<std::shared_ptr<Anchor>>& anchors) = 0;
```
this is a tab

lincolnfrog commented on third_party/tango/libtango_chromium/TangoHandler.h:
```diff
@@ -77,7 +82,8 @@ class TangoHandler {
 
-  void onCreate(JNIEnv* env, jobject activity, int activityOrientation, int sensorOrientation);
+  void onCreate(JNIEnv* env, jobject activity, 
+  							int activityOrientation, int sensorOrientation);
```
these are tabs

lincolnfrog commented on third_party/tango/libtango_chromium/TangoHandler.h:
```diff
@@ -89,9 +95,14 @@ class TangoHandler {
   bool hitTest(float x, float y, std::vector<Hit>& hits);
   bool getPlaneDeltas(PlaneDeltas& planeDeltas);
+  std::shared_ptr<Anchor> createAnchor(
+  	const float* anchorModelMatrix);
```
this is a tab - I don't think this even needs to be two lines

lincolnfrog reviewed (changes requested):
Lots of style issues. I think we should fix but up to you how strict we want to be

judax committed:
Fixed the style issues to adhere to Google CStyle guidelines.

Change-Id: Id45928969bdec0f35ebb5b68db0f23ec690f7a3b

judax commented:
Added a patch to fix these styling issues.

judax committed:
Some major changes. More modular structure: separated Anchor and
AnchorManager, created the new headers for Math and Log Utils. Fixed the
anchor updating algorithm was incorrect as the device pose was being
calculate for the history change timestamp and it has to be calculated
for each anchor timestamp specifically.

Change-Id: I220a5742e178c34373a8e631655b4f00c7adc70e
